lib/vm/src/instance/ref.rs

```diff
@@ -155,3 +155,75 @@ impl InstanceRef {
         (&mut *ptr).as_mut()
     }
 }
+
+/// A weak instance ref. This type does not keep the underlying `Instance` alive
+/// but can be converted into a full `InstanceRef` if the underlying `Instance` hasn't
+/// been deallocated.
+#[derive(Debug, Clone)]
+pub struct WeakInstanceRef(Weak<InstanceInner>);
+
+impl PartialEq for WeakInstanceRef {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.ptr_eq(&other.0)
+    }
+}
+
+impl WeakInstanceRef {
+    /// Try to convert into a strong, `InstanceRef`.
+    pub fn upgrade(&self) -> Option<InstanceRef> {
+        let inner = self.0.upgrade()?;
+        Some(InstanceRef(inner))
+    }
+}
+
+impl MemoryUsage for WeakInstanceRef {
+    fn size_of_val(&self, tracker: &mut dyn MemoryUsageTracker) -> usize {
+        mem::size_of_val(self)
+            + if let Some(ir) = self.upgrade() {
```

Comment:
Getting the size of a value makes it strong? That seems wrong.

Reply:
Yeah, weak references can't be accessed. The only way to look into this type and see how much memory might be   being used is to try to make it into a strong reference. If it can become a strong reference then that memory exists and we can count it, if it's not then we can't.

Reply:
And to clarify it doesn't mutate the internal value, it just needs to upgrade the reference in order to count correctly.  This logic should be implemented inside of loupe itself on the `Weak` type, not here.